File: LinkedList/SplitLLInParts/split.cpp
#include <iostream>
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        ListNode* temp = head;
        int n = 0;
        while(temp!=NULL){
            n++;
            temp=temp->next;
        }
        int rem=n%k;
        vector<ListNode*>ans;
        temp = head;
        while(temp!=NULL){
            ListNode* tempC = new ListNode(100);
            ListNode* c = tempC;
            int size=n/k;
            if(rem>0)size++;
            rem--;
            for(int i=1;i<=size;i++){
                tempC->next=temp;
                temp=temp->next;
                tempC=tempC->next;
            }
            tempC->next=NULL;
            ans.push_back(c->next);
        }
        if(ans.size()<k)while(ans.size()<k)ans.push_back(NULL);
        return ans;
    }
};

File: LinkedList/SplitLLInParts/split.py
from typing import Optional,List

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        if k==1: return [head]
        n = 0
        temp = head
        while temp:
            n += 1
            temp =temp.next
        totalSize = n//k
        extraSize = n%k
        ans = [None]*k
        temp = head
        listNum = 0
        while temp:
            sizeOfLL = totalSize
            if extraSize:
                sizeOfLL += 1
                extraSize -= 1
            count = 1
            newHead = temp
            while count < sizeOfLL:
                temp = temp.next
                count += 1
            t = temp
            temp = temp.next
            t.next = None
            ans[listNum] = newHead
            listNum += 1
        return ans